src/NodaTime/TimeInterval.cs

```diff
@@ -0,0 +1,284 @@
+    /// may be missing its start or end, in which case the TimeInterval is deemed to be infinite in that
+    /// direction.
+    /// </para>
+    /// <para>
+    /// The end may equal the start (resulting in an empty TimeInterval),
+    /// but it can also be before the start: in this case property IsEndInNextDay will be true
+    /// </para>
+    /// </remarks>
+    /// <threadsafety>This type is an immutable value type. See the thread safety section of the user guide for more information.</threadsafety>
+    public readonly struct TimeInterval : IEquatable<TimeInterval>
+    {
+        /// <summary>The start of the interval.</summary>
+        private readonly LocalTime? start;
+
+        /// <summary>The end of the interval. It can be less than start.</summary>
+        private readonly LocalTime? end;
+        
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TimeInterval"/> struct.
+        /// The TimeInterval includes the <paramref name="start"/> LocalTime and excludes the
+        /// <paramref name="end"/> LocalTime. The end may equal the start (resulting in an empty TimeInterval).
+        /// </summary>
+        /// <param name="start">The start <see cref="LocalTime"/>.</param>
+        /// <param name="end">The end <see cref="LocalTime"/>.</param>
+        public TimeInterval(LocalTime start, LocalTime end)
+        {
+            this.start = start;
+            this.end = end;
+            IsEndInNextDay = start > end;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TimeInterval"/> struct from two nullable <see cref="LocalTime"/>
+        /// values.
+        /// </summary>
+        /// <remarks>
+        /// If the start is null, the TimeInterval is deemed to stretch to the start of time. If the end is null,
+        /// the TimeInterval is deemed to stretch to the end of time.
+        /// </remarks>
+        /// <param name="start">The start <see cref="LocalTime"/> or null.</param>
+        /// <param name="end">The end <see cref="LocalTime"/> or null.</param>
+        public TimeInterval(LocalTime? start, LocalTime? end)
+        {
+            this.start = start;
+            this.end = end;
+            if (start.HasValue && end.HasValue)
+                IsEndInNextDay = start > end;
+            else
+                IsEndInNextDay = false;
+        }
+
+        /// <summary>
+        /// Gets the start LocalTime - the inclusive lower bound of the TimeInterval.
+        /// </summary>
+        /// <value>The start <see cref="LocalTime"/>.</value>
+        /// <exception cref="InvalidOperationException">The TimeInterval extends to the start of time.</exception>
+        /// <seealso cref="HasStart"/>
+        public LocalTime Start
+        {
+            get
+            {
+                Preconditions.CheckState(start.HasValue, "TimeInterval extends to start of time");
```

Comment:
One advantage of not having a nullable start is that we don't need to worry about this.